File: Data Structures & Algorithms/interleaving-string/submission-3.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool: 
        cache = set()
 
        def dfs(i1, i2):
            i3 = i1+i2
            if i3 == len(s3):
                return (i1 == len(s1)) and (i2 == len(s2))
            if (i1, i2) in cache:
                return False

            if i1 < len(s1) and s1[i1] == s3[i3]:
                if dfs(i1 + 1, i2):
                    return True
            
            if i2 < len(s2) and s2[i2] == s3[i3]:
                if dfs(i1, i2 + 1):
                    return True
            
            cache.add((i1, i2))
            return False
                    
    
        return dfs(0, 0)
        

                    


                

File: Data Structures & Algorithms/course-schedule-ii/submission-3.ts
class Solution {
    /**
     * @param {number} numCourses
     * @param {number[][]} prerequisites
     * @return {number[]}
     */
    findOrder(numCourses: number, prerequisites: number[][]): number[] {
    // Step 1: Build Adjacency List
    const adj: number[][] = Array.from({ length: numCourses }, () => []);
    for (const [course, prereq] of prerequisites) {
        adj[prereq].push(course); // prereq -> course
    }

    // State array: 0 = Unvisited, 1 = Visiting, 2 = Visited
    const state: number[] = new Array(numCourses).fill(0);
    const result: number[] = [];

    // Step 2: DFS Helper Function
    function hasCycle(node: number): boolean {
        state[node] = 1; // Mark as VISITING (enter call stack)

        for (const neighbor of adj[node]) {
            // Cycle condition: encountered a node currently in active stack
            if (state[neighbor] === 1) return true;
            
            // Unvisited node: recursively visit
            if (state[neighbor] === 0 && hasCycle(neighbor)) return true;
        }

        state[node] = 2;   // Mark as VISITED (exit call stack)
        result.push(node); // Post-order record
        return false;
    }

    // Step 3: Run DFS from every unvisited node
    for (let i = 0; i < numCourses; i++) {
        if (state[i] === 0) {
            if (hasCycle(i)) return []; // Return empty array if any cycle found
        }
    }

    // Step 4: Reverse post-order result to get true topological order
    return result.reverse();
}
}
